### PR TITLE
Improve the error calculation in MonteCarloAbsorption

### LINKED ISSUE #28932: Investigate improving error calculation on Monte Carlo absorption corrections
The error on the absorption correction values output from MonteCarloAbsorption is currently set to 1/sqrt(nevents) in the constructor MCAbsorptionStrategy.cpp
Investigate whether we can implement a better estimate of the error (as discussed with instrument scientists in absorption correction presentation on Thurs 25th June). Should it instead be the standard deviation of the "per track" absorption factors divided by sqrt(nevents)?

### PR DESCRIPTION
**Pull request by DannyHindson**
**Description of work.**

The error on each point in the attenuation factor spectrum was previously calculated as 1/sqrt(nevents). This error was the same for all points. This has been changed to the standard deviation/sqrt(nevents) which is the appropriate calculation given that each point in the spectrum is a mean of the attenuation factors calculated over various simulated tracks through the sample

At this stage, the error is only output if you're not using any of the options that require interpolation:
a) wavelength interpolation. This is only used if the ResimulateTracksForDifferentWavelengths parameter is set to True which is not the default. The gsl routines for interpolation don't appear to return interpolation errors although seems like we only use gsl for spline interpolation. May require us to implement our own spline interpolation 
b) using a a sparse instrument. The spatial interpolation used in the sparse calculation means the errors need to be propagated through the interpolation calculation which needs further work

I've added some unit tests to check the errors in a couple of specific situations where it's possible to do an exact back of an envelope calculation of the mean\standard deviation of the absorption factor across all possible paths through the shape. The unit tests for the classes MonteCarloAbsorption, MCAbsorptionStrategy, MCInteractionVolume overlap to some extent at the moment because of the dependencies between them so will also create a separate issue to add in some mocking to make it a bit cleaner

**To test:**

Run an absorption correction calculation on a non-sparse instrument such as the following one that is based on one of the usage examples from the MonteCarloAbsorption help page:
```
data = CreateSampleWorkspace(WorkspaceType='Histogram', NumBanks=1)
data = ConvertUnits(data, Target="Wavelength")
# Default up axis is Y
SetSample(data, Geometry={'Shape': 'Cylinder', 'Height': 5.0, 'Radius': 1.0,
                  'Center': [0.0,0.0,0.0]},
                Material={'ChemicalFormula': '(Li7)2-C-H4-N-Cl6', 'SampleNumberDensity': 0.07})
# Simulating every data point can be slow. Use a smaller set and interpolate
abscor = MonteCarloAbsorption(InputWorkspace=data,EventsPerPoint=1000)
corrected = data/abscor
```
Look at the error values in the abscor workspace and note that they are now different for all the bins\spectra
Run the script again with more scenarios (EventsPerPoint=10000) and note that the error has gone down

Fixes #28932 .

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by cailafinn — CHANGES_REQUESTED** [review]
Code looks good. Did notice while testing that if the radius of the sample is set to 5, increasing the `EventsPerPoint` caused the error to go up in some spectra. Is this to be expected?

**Comment by DannyHindson** [comment]
@ConorMFinn - I had a look into your question and it's highlighted something I've not fully appreciated before. The absorption correction values are an average over lots of simulated tracks and the result is affected a lot by tracks which only skim the surface of the sample: 
CylinderSampleDiagram
The attenuation factor varies exponentially with the distance the tracks pass through so that means these "skim" tracks contribute a lot to the mean attenuation factor (and their effect on the error is even greater). When run with a relatively small number of events (eg 1000) you only get a few of these skim tracks, whereas when run on a bigger sample you get a more representative quantity of these.
So in summary it means the current calculation underestimates the error when run on a small number of events so my original assertion that the error will reduce as you increase the number of tracks only kicks in once the number of events is above a certain number.
I think this underestimation could be corrected for by using a more complicated error calculation than the classic "standard deviation" but probably beyond the scope of this PR.
It's possible this effect is amplified when running on bigger samples. Not done the maths to show this but I'm guessing that's why it showed up when you increased the sample radius.

**Review by cailafinn — APPROVED** [review]
Concern addressed. Agree that an more complex error calculation is out of the scope of the PR as for the majority of cases the error calculation has been improved. Thanks for clarifying that for me. Nice work 👍

File: docs/source/algorithms/MonteCarloAbsorption-v1.rst [hunk 1]
```diff
@@ -112,6 +112,7 @@ When the sparse instrument option is enabled, a sparse instrument corresponding
 The interpolation is a two step process: first a spatial interpolation is done from the detector grid of the sparse instrument to the actual detector positions of the full instrument. Then, if `ResimulateTracksForDifferentWavelengths` = True the correction factors are interpolated over the missing wavelengths.
 
 .. note:: Currently, the sparse instrument mode does not support instruments with varying *EFixed*.
+.. note:: Errors are not currently provided for absorption correction factors calculated involving spatial interpolation
```
Comment (gvardany) [inline docs/source/algorithms/MonteCarloAbsorption-v1.rst:115]: Is there a fundamental reason for this?
Would it makes sense to interpolate also the errors?

Comment (DannyHindson) [reply docs/source/algorithms/MonteCarloAbsorption-v1.rst:115]: I think we should extend this to calculate the errors for interpolated points also. The only reason I've not done it here was that it seemed a relatively large piece of work so I've created another issue for it (#29000). I think the error calculation for the interpolated points would include two contributions: 
a) propagate errors from the simulated points
b) include a contribution from the "interpolation error" ie how well does the interpolation function match the actual one

Comment (gvardany) [reply docs/source/algorithms/MonteCarloAbsorption-v1.rst:115]: My guess is that b) could be ignored if you have high order splines. The corrections are supposed to spatially smooth (except for the flat plate at the sample angle), but there correction is problematic anyhow, so one should be careful with using spare instrument in that case. Agreed that a) can be done in a separate PR and the gross approximation would be just to run the same spline interpolation on the 'errors'.

Comment (DannyHindson) [reply docs/source/algorithms/MonteCarloAbsorption-v1.rst:115]: OK good. It was estimating the interpolation error that seemed the hardest part. I was hoping that the gsl library we use for the spline interpolation returned the interpolation error but surprisingly it doesn't.

Comment (gvardany) [reply docs/source/algorithms/MonteCarloAbsorption-v1.rst:115]: I knew it doesn't since the exact same problem is there for SplineInterpolation algorithm; no errors at the moment.
